File: sorento_crm_backend/app/services/scm/plan_grain.py
"""Which grain a front-planning run may be decided at, and who may write to it.

Plan grain (front-planning plan section 5.1) is an ADMIN POLICY, not a per-run selector
the buyer picks: one field in system settings, rollout default `product`. A run stamps the
configured grain at creation (`reorder_run.decision_grain`) together with the contract
version, and neither ever changes afterwards - so changing the policy affects only runs
created later, and the grain a screen shows for an existing run is its stamp, never the
live setting (AC-F09, AC-F10).

Two writes are therefore refused, and both are 409 because the request is well-formed and
the state is what makes it wrong:

* `legacy_run_read_only` - the run predates the contract (`front_planning_contract_version
  IS NULL`). Its recommendations, overrides and summary rows stay exactly as they are, and
  the buyer creates a new run to act. `decision_grain` is deliberately NOT backfilled;
  guessing which grain a historical run "meant" is the one thing that could make an old
  decision actionable twice.
* `decision_grain_mismatch` - the run is stamped at the OTHER grain. Product decisions are
  the `order_summary_row` chosen quantity; location decisions are the recommendation
  accept / adjust / reject. Exactly one of the two is actionable per run, which is what
  stops both grains ordering the same requirement.

Both services route through `assert_decision_grain` so the two refusals cannot drift apart
in wording or in status code.
"""
from __future__ import annotations

from typing import Optional

from sqlalchemy.orm import Session

from app.services.error_handler import AppException

PRODUCT_GRAIN = "product"
LOCATION_GRAIN = "location"
PLAN_GRAINS = (PRODUCT_GRAIN, LOCATION_GRAIN)

#: The rollout default (plan 5.1). Also what an unset / unreadable setting resolves to.
DEFAULT_PLAN_GRAIN = PRODUCT_GRAIN

#: Stamped on every run created under this contract. Bumped only if the front-planning
#: contract itself changes shape; NULL means "before the contract", i.e. legacy.
FRONT_PLANNING_CONTRACT_VERSION = 1

_GRAIN_LABEL = {
    PRODUCT_GRAIN: "Product",
    LOCATION_GRAIN: "Location",
}


def resolve_plan_grain(db: Session) -> str:
    """The configured admin policy, or the rollout default when nothing is configured."""
    from app.models.user import SystemSetting

    value = db.query(SystemSetting.plan_grain).limit(1).scalar()
    return value if value in PLAN_GRAINS else DEFAULT_PLAN_GRAIN


def is_legacy_run(run) -> bool:
    """A run created before the front-planning contract. Read-only, both grains."""
    return getattr(run, "front_planning_contract_version", None) is None


def decision_grain_of(run) -> Optional[str]:
    """The run's stamped grain, or None for a legacy run."""
    if is_legacy_run(run):
        return None
    return getattr(run, "decision_grain", None)


def assert_decision_grain(run, expected: str) -> None:
    """Refuse a decision write this run cannot own. Raises `AppException` (409)."""
    if is_legacy_run(run):
        raise AppException(
            status_code=409,
            message=(
                "This plan was created before product-and-location planning, so it is "
                "read only. Create a new plan to decide on it."
            ),
            code="legacy_run_read_only",
        )
    grain = getattr(run, "decision_grain", None)
    if grain != expected:
        raise AppException(
            status_code=409,
            message=(
                f"This plan is decided at {_GRAIN_LABEL.get(grain, grain)} grain, so a "
                f"{_GRAIN_LABEL.get(expected, expected)} decision cannot be recorded on "
                "it."
            ),
            code="decision_grain_mismatch",
        )

File: sorento_crm_frontend/app/(protected)/scm/reorder/lib/planGrain.ts
/**
 * Plan grain - the run's stamped decision grain, read by every planning surface.
 *
 * Plan grain is ADMIN POLICY (plan section 5.1). It is not a per-run selector and
 * there is deliberately no control anywhere on the plan page that changes it: a run
 * stamps the configured grain when it is created and keeps it forever, so what the
 * buyer sees here is a fact about the run in front of them, never a choice.
 *
 * Three states, and the third is not a value of the first two:
 *
 * - `product` - the single product row owns the chosen quantity; the per-location
 *                  view stays open as a read and drill view (AC-F02).
 * - `location` - the per-location recommendations and overrides stay actionable; the
 *                  product row is a read-only aggregate of them.
 * - LEGACY   - a run created before the front-planning contract. It carries no
 *                  grain and no channel breakdown, accepts no decision in either grain,
 *                  and is never backfilled (AC-F10). A legacy run is identified by its
 *                  NULL contract version, not by a missing grain alone, because that is
 *                  the field the backend rejects writes on.
 */

export type PlanGrain = 'product' | 'location';

/** `decisionLockReason`'s own wording for the Product-grain case, exported so a caller that
 *  wants to say MORE than the flat sentence (the Order summary sheet's own product-grain
 *  lock, `SummaryOrderReportView.tsx`) can recognise it without duplicating the string. NOT
 *  used by the reorder plan row any more (S16, 21 Aug) - see `legacyLockReason` below. */
export const DECIDED_AT_PRODUCT_GRAIN = 'Decided at Product grain';

/** Shared wording for a run that predates the front-planning contract - its decisions are
 *  history, and nothing here may rewrite them. */
const LEGACY_LOCK_MESSAGE = 'Legacy run - read only. Create a new plan to decide.';

/** The two run fields every grain question is answered from. */
export interface RunGrainState {
  /** Stamped once at run creation. NULL only on a legacy run. */
  decision_grain?: PlanGrain | null;
  /** `1` on a front-planning run, NULL on a legacy one. */
  front_planning_contract_version?: number | null;
}

/** A run that predates the front-planning contract: read-only in both grains. */
export function isLegacyRun(run: RunGrainState | null | undefined): boolean {
  if (!run) return false;
  return (
    run.front_planning_contract_version === null ||
    run.front_planning_contract_version === undefined ||
    !run.decision_grain
  );
}

/** What the header chip says. Legacy is its own word, not an absent grain. */
export function planGrainLabel(run: RunGrainState | null | undefined): string {
  if (!run) return 'Plan grain unknown';
  if (isLegacyRun(run)) return 'Legacy run';
  return run.decision_grain === 'location' ? 'Plan grain: Location' : 'Plan grain: Product';
}

/**
 * Whether the Buy view groups its per-warehouse rows into one row per (product, channel)
 * (5.3: "1 line of retail, 1 line of project"). Product-grain runs only: under Location
 * grain the per-warehouse row IS the decision surface (5.4), so grouping it away would
 * hide the thing the buyer is deciding over. A legacy run has no stamped grain at all.
 */
export function shouldGroupByChannel(run: RunGrainState | null | undefined): boolean {
  if (!run || isLegacyRun(run)) return false;
  return run.decision_grain === 'product';
}

/**
 * Why a decision control is disabled, or null when it is actionable.
 *
 * `surface` is the grain the control would write in. A run answers exactly one of
 * them, so the other surface stays open for reading and says which screen owns the
 * decision instead of silently doing nothing.
 */
export function decisionLockReason(
  run: RunGrainState | null | undefined,
  surface: PlanGrain,
): string | null {
  if (!run) return null;
  if (isLegacyRun(run)) return LEGACY_LOCK_MESSAGE;
  if (run.decision_grain === surface) return null;
  return run.decision_grain === 'product'
    ? DECIDED_AT_PRODUCT_GRAIN
    : 'Decided at Location grain';
}

/**
 * Why the reorder plan row's OWN decision control is locked (S16, captain 21 Aug, 3rd
 * time requested: "I want the decision made here"). Unlike `decisionLockReason` above -
 * still the Order summary sheet's own lock, and still genuinely grain-scoped there - the
 * plan row is a decision surface at EVERY grain now, grouped included (a product-grain
 * row fans the same decision out to its members). The only run this can never write for
 * is one that predates the front-planning contract, whose decisions are history.
 */
export function legacyLockReason(run: RunGrainState | null | undefined): string | null {
  if (!run) return null;
  return isLegacyRun(run) ? LEGACY_LOCK_MESSAGE : null;
}
